src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -0,0 +1,264 @@
+                // If we have seen this page before, we just continue down the path. 
+                if (_pagesSeen.Contains(state.Page.PageNumber))
+                {
+                    var next = GetValue(ref state, _generator.Next(state.Header->NumberOfEntries));
+                    _tree.PushPage(next, ref cState);
+
+                    state = ref cState._stk[cState._pos];
+                }
+                else
+                {
+                    // We havent seen it, so we are going to be processing it. 
+
+                    // If the current index is already out of bounds, we are going to be marking this page as seen and move
+                    // to the next random page down the path.
+                    if (_currentIdx >= state.Header->NumberOfEntries)
+                    {
+                        _pagesSeen.Add(state.Page.PageNumber);
+                        _currentIdx = 0;
+
+                        // Push the next random page.
+                        var next = GetValue(ref state, _generator.Next(state.Header->NumberOfEntries));
+                        _tree.PushPage(next, ref cState);
+                        state = ref cState._stk[cState._pos];
+                    }
+                    else
+                    {
+                        GetEntry(_tree, state.Page, state.EntriesOffsets[_currentIdx], out keySpan, out var value);
+                        _currentIdx++;
+                        goto Success;
+                    }
+                }
+            }
+
+            // If the current page is a leaf we are going to retrieve a single random key and setup everything
+            // to start the random process of keys generation from the root again on the next call. 
+            if (state.Header->PageFlags.HasFlag(CompactPageFlags.Leaf))
+            {
+                int randomEntry = _generator.Next(state.Header->NumberOfEntries);
+                GetEntry(_tree, state.Page, state.EntriesOffsets[randomEntry], out keySpan, out var value);
+
+                // Move the cursor to the root.                 
+                _cursor._pos = -1;
+                _cursor._len = 0;
+                _tree.PushPage(_tree._state.RootPage, ref _cursor);
+                _currentIdx = 0;
+            }
+
+            Debug.Assert(keySpan.Length != 0, "If no key has been retrieved, there is an error in the algorithm.");
+
+            Success:
+            key = keySpan;
+            _currentSample++;
+            return true;
+
+            Failure:
+            key = Span<byte>.Empty;
+            return false;
+        }
+
+        public void Reset()
+        {
+            _cursor._pos = -1;
+            _cursor._len = 0;
+            _tree.PushPage(_tree._state.RootPage, ref _cursor);
+
+            _currentSample = 0;
+            _currentIdx = 0;
+        }
+    }
+
+    public bool ShouldImproveDictionary()
+    {
+        if (_llt.Flags != TransactionFlags.ReadWrite)
+            return false;
```

Comment:
Why do we have this check here?

Reply:
Because it shouldn't be called, it is either ignore OR throw an exception. We simply cannot allow it to continue to run a dictionary improve. In a sense this is more like a try condition. Returning `false` is safe and semantically correct. If we dont check it will blow up eventually when we request write access.

Reply:
This is already checked by the caller (using `Debug.Assert`).
Returning true here means that you are modifying the tree, which will error if not in write tx anyway.